xla/service/gpu/runtime/copy_thunk.cc

```diff
@@ -45,40 +46,108 @@ absl::Status DeviceToDeviceCopyThunk::ExecuteOnStream(
+// Emplace() will insert {key, event} pair into the hash map,
+// and return the event in order to do RecordEvent() for async memcpy.
+absl::Status CopyAsyncEvents::Emplace(se::StreamExecutor *executor,
+                                      const HloInstruction *instr,
+                                      se::Event &&event) {
+  Key key = {executor, instr};
+  absl::MutexLock lock(&mutex_);
+  VLOG(3) << "Emplace event " << event.implementation();
+  if (auto [it, inserted] = events_.try_emplace(key, std::move(event));
+      inserted) {
+    return absl::OkStatus();
+  }
+  return absl::InternalError("Async copy event already exists!");
 }
 
-HostToDeviceCopyThunk::HostToDeviceCopyThunk(
-    ThunkInfo thunk_info, const BufferAllocation::Slice& source_buffer,
-    const BufferAllocation::Slice& destination_buffer, uint64_t mem_size)
+// Retrieve a completion event started by copy-start instruction
+// `instr`, and remove the event from the collection.
+absl::StatusOr<se::Event>
+CopyAsyncEvents::Extract(se::StreamExecutor *executor,
+                         const HloInstruction *instr) {
+  Key key = {executor, instr};
+  absl::MutexLock lock(&mutex_);
+  if (auto event = events_.extract(key)) {
+    VLOG(3) << "Extract event " << event.mapped().implementation();
+    return std::move(event.mapped());
+  }
+  return absl::InternalError("Async copy event was not found!");
+}
+
+//===----------------------------------------------------------------------===//
+// DeviceHostCopyThunk
+//===----------------------------------------------------------------------===//
+DeviceHostCopyThunk::DeviceHostCopyThunk(
+    ThunkInfo thunk_info, const BufferAllocation::Slice &source_buffer,
+    const BufferAllocation::Slice &destination_buffer, uint64_t mem_size,
+    std::shared_ptr<CopyAsyncEvents> async_events, const HloInstruction *instr,
+    CopyDirection copy_direction)
     : DeviceToDeviceCopyThunk(thunk_info, source_buffer, destination_buffer,
-                              mem_size) {}
+                              mem_size),
+      async_events_(std::move(async_events)), instr_(instr),
+      copy_direction_(copy_direction) {}
 
-absl::Status HostToDeviceCopyThunk::ExecuteOnStream(
+absl::Status DeviceHostCopyThunk::ExecuteOnStream(
     const ExecuteParams& params) {
   se::DeviceMemoryBase destination_data =
       params.buffer_allocations->GetDeviceAddress(destination());
   se::DeviceMemoryBase source_data =
       params.buffer_allocations->GetDeviceAddress(source());
-  void* cpu_src = source_data.opaque();
-  VLOG(3) << "Memcpy H2D for memory offload from " << cpu_src << " to "
-          << destination_data.opaque();
-  return params.stream->Memcpy(&destination_data, cpu_src, size_bytes());
+  void *cpu_dst = destination_data.opaque();
+  void *cpu_src = source_data.opaque();
+  TF_ASSIGN_OR_RETURN(
+      se::Stream * stream,
+      GetStreamForExecution(Thunk::execution_stream_id(), params));
+  std::string vlog_string = "Memcpy ";
```

Comment:
I think single absl::StrtFormat(...) will be much simpler to read then following all updates to vlog_string

Reply:
Removed vlog_string and use VLOG() directly.